linkerd/system/src/linux.rs

```diff
@@ -39,9 +42,28 @@ pub fn open_fds(pid: pid_t) -> io::Result<u64> {
 }
 
 pub fn max_fds() -> io::Result<Option<u64>> {
-    let limit = pid::limits_self()?.max_open_files;
-    let max_fds = limit.soft.or(limit.hard).map(|max| max as u64);
-    Ok(max_fds)
+    match Process::myself() {
+        Ok(p) => match p.limits() {
+            Ok(limits) => {
+                let limit = limits.max_open_files;
+                match limit.soft_limit {
+                    process::LimitValue::Unlimited => match limit.hard_limit {
+                        process::LimitValue::Unlimited => Ok(None),
+                        process::LimitValue::Value(v) => Ok(Some(v)),
+                    },
+                    process::LimitValue::Value(v) => Ok(Some(v)),
+                }
+            }
+            Err(e) => {
+                error!("Failed to get process limits: {}", e);
+                Ok(None)
```

Comment:
It should be fine because this is how we use the max_fds
```rust
match sys::max_fds() {
                Ok(None) => {}
                Ok(Some(max_fds)) => {
                    process_max_fds.fmt_help(f)?;
                    process_max_fds.fmt_metric(f, &max_fds.into())?;
                }
                Err(err) => {
                    warn!("Could not determine process_max_fds: {}", err);
                }
            }
```

Also, in the previous implementation we were not really handling the errors, instead we ended up assuming that hard limit will always exist